pkg/operator/ceph/object/shared_pools.go

```diff
@@ -0,0 +1,695 @@
+	if err != nil {
+		return errors.Wrapf(err, "failed to list pools")
+	}
+	existing := make(map[string]struct{}, len(existingPools))
+	for _, pool := range existingPools {
+		existing[pool.Name] = struct{}{}
+	}
+	// ignore pools with empty name:
+	existing[""] = struct{}{}
+
+	if _, ok := existing[sharedPools.MetadataPoolName]; !ok {
+		return fmt.Errorf("sharedPool do not exist: %s", sharedPools.MetadataPoolName)
+	}
+	if _, ok := existing[sharedPools.DataPoolName]; !ok {
+		return fmt.Errorf("sharedPool do not exist: %s", sharedPools.DataPoolName)
+	}
+
+	for _, pp := range sharedPools.PoolPlacements {
+		if _, ok := existing[pp.MetadataPoolName]; !ok {
+			return fmt.Errorf("sharedPool do not exist: pool %s for placement %s", pp.MetadataPoolName, pp.Name)
+		}
+		if _, ok := existing[pp.DataPoolName]; !ok {
+			return fmt.Errorf("sharedPool do not exist: pool %s for placement %s", pp.DataPoolName, pp.Name)
+		}
+		if _, ok := existing[pp.DataNonECPoolName]; !ok {
+			return fmt.Errorf("sharedPool do not exist: pool %s for placement %s", pp.DataNonECPoolName, pp.Name)
+		}
+		for _, sc := range pp.StorageClasses {
+			if _, ok := existing[sc.DataPoolName]; !ok {
+				return fmt.Errorf("sharedPool do not exist: pool %s for StorageClass %s", sc.DataPoolName, sc.Name)
+			}
+		}
+	}
+
+	logger.Info("verified shared pools exist")
+	return nil
+}
+
+func ConfigureSharedPoolsForZone(objContext *Context, sharedPools cephv1.ObjectSharedPoolsSpec) error {
+	if sharedPools.DataPoolName == "" && sharedPools.MetadataPoolName == "" && len(sharedPools.PoolPlacements) == 0 {
+		logger.Debugf("no shared pools to configure for store %q", objContext.Name)
+		return nil
+	}
+
+	if err := sharedPoolsExist(objContext, sharedPools); err != nil {
+		return errors.Wrapf(err, "object store cannot be configured until shared pools exist")
+	}
+
+	logger.Infof("Retrieving zone %q", objContext.Zone)
+	zoneConfig, err := getZoneJSON(objContext)
+	if err != nil {
+		return err
+	}
+	logger.Infof("Updating rados namespace configuration for zone %q", objContext.Zone)
+	zoneChanged, err := adjustZoneDefaultPools(zoneConfig, sharedPools)
+	if err != nil {
+		return err
+	}
+	changed, err := adjustZonePlacementPools(zoneConfig, sharedPools)
+	if err != nil {
+		return err
+	}
+	zoneChanged = zoneChanged || changed
+
+	logger.Infof("Retrieving zonegroup %q", objContext.Zone)
+	zoneGroupConfig, err := getZoneGroupJSON(objContext)
+	if err != nil {
+		return err
+	}
+	logger.Infof("Updating rados namespace configuration for zonegroup %q", objContext.ZoneGroup)
+	placements, err := zoneConfig.getPoolPlacements()
+	if err != nil {
+		return err
+	}
+	zoneGroupChanged, err := adjustZoneGroupPlacementTargets(zoneGroupConfig, placements)
+	if err != nil {
+		return err
+	}
+
+	// persist configuration updates:
+	if zoneChanged {
+		logger.Infof("zone config changed: performing zone config updates for %s", objContext.Zone)
```

Comment:
Any `logger.X()` methods must include the object store (or zone) name and namespace (in namespaced-name format) so that we can disambiguate these log lines when there are multiple object stores and/or ceph clusters. Please check all `logger.X()` usage, not just this line.

This is theoretically implied when we see a zone update command in rook logs, but debugging these kinds of branches can be tricky. Let's **keep** this info log under the `if zoneChanged` conditional as a clarification.